File: C2_bessel.py
import numpy as np
import matplotlib.pyplot as plt

def f(t,x,n):
    return np.cos(n*t - x*np.sin(t))

def simpson(a,b,nsteps,x,n):             #n es el número de intervalos (debe ser entero y par)
    nsteps *= 2
    h = (b-a)/nsteps                 #es el paso que damos 
    sum = 0

    for i in range(nsteps+1):        #n+1 para que comience en 0 y termine en n (necesito el dato n-1)
        t = a + i*h
        if(i==0 or i==nsteps):
            sum += f(t,x,n)
        elif(i%2 == 0):
            sum += 2*f(t,x,n)
        else:
            sum += 4*f(t,x,n)
    
    return sum*h/3

def bessel(n,x):
    return (1/np.pi)*simpson(0,np.pi,100,x,n)


#GRAFICAR

n = 0
a = 0
b = 10
Nsteps = 100
dx = (b-a)/Nsteps
xdata = np.zeros(Nsteps)
ydata = np.zeros(Nsteps)

for i in range(Nsteps):
    x = a + i*dx
    xdata[i] = x
    ydata[i] = bessel(n,x)

plt.plot(xdata, ydata)
plt.show()

File: C2_bessel.cpp
#include <iostream>
#include <cmath>
using namespace std;

double f(double t, double x, double n){
    return cos(n*t - x*sin(t));
}

double simpson(double a, double b, int nsteps, double x, double n){
    nsteps *= 2;
    double dx = (b-a)/nsteps;
    double sum = 0;

    for(int i = 0; i < nsteps + 1; i++){
        double t = a + i*dx;
        if(i==0 || i==nsteps){
            sum += f(t,x,n);
        }
        else if(i%2 == 0){
            sum += 2*f(t,x,n);
        }
        else{
            sum += 4*f(t,x,n);
        }
    }
    return sum*dx/3;
    }

double bessel(double n, double x){
    return (1/M_PI)*simpson(0,M_PI,100,x,n);
}

int main(){

    double n = 0;
    double a = 0;
    double b = 10;
    double nsteps = 100;
    double dx = (b-a)/nsteps;

    for(int i = 0; i < nsteps; i++){
        double x = a + i*dx;

        cout << x << " " << bessel(n,x) <<endl;
    }

    return 0;
}